File: app/src/main/java/com/cargopull/executor_driver/di/AppComponent.java
package com.cargopull.executor_driver.di;

import com.cargopull.executor_driver.application.BaseActivity;
import com.cargopull.executor_driver.application.FcmService;
import com.cargopull.executor_driver.application.MainApplication;
import com.cargopull.executor_driver.application.MenuActivity;
import com.cargopull.executor_driver.application.MovingToClientActivity;
import com.cargopull.executor_driver.application.NightModeActivity;
import com.cargopull.executor_driver.application.OnlineActivity;
import com.cargopull.executor_driver.application.OnlineMenuActivity;
import com.cargopull.executor_driver.application.OrderCostDetailsActivity;
import com.cargopull.executor_driver.application.OrderFulfillmentActivity;
import com.cargopull.executor_driver.application.PasswordActivity;
import com.cargopull.executor_driver.application.PreOrdersActivity;
import com.cargopull.executor_driver.application.WaitingForClientActivity;
import com.cargopull.executor_driver.backend.settings.AppSettingsService;
import com.cargopull.executor_driver.utils.Consumer;
import com.cargopull.executor_driver.view.ActionOrderDetailsFragment;
import com.cargopull.executor_driver.view.BalanceFragment;
import com.cargopull.executor_driver.view.BalanceSummaryFragment;
import com.cargopull.executor_driver.view.CallToClientFragment;
import com.cargopull.executor_driver.view.CallToOperatorFragment;
import com.cargopull.executor_driver.view.CancelOrderDialogFragment;
import com.cargopull.executor_driver.view.ChooseVehicleFragment;
import com.cargopull.executor_driver.view.ClientOrderConfirmationFragment;
import com.cargopull.executor_driver.view.ClientOrderConfirmationTimeFragment;
import com.cargopull.executor_driver.view.DriverOrderConfirmationFragment;
import com.cargopull.executor_driver.view.DriverPreOrderBookingFragment;
import com.cargopull.executor_driver.view.DriverPreOrderConfirmationFragment;
import com.cargopull.executor_driver.view.GeoEngagementDialogFragment;
import com.cargopull.executor_driver.view.GoOnlineFragment;
import com.cargopull.executor_driver.view.MapFragment;
import com.cargopull.executor_driver.view.MenuFragment;
import com.cargopull.executor_driver.view.MovingToClientActionsDialogFragment;
import com.cargopull.executor_driver.view.MovingToClientFragment;
import com.cargopull.executor_driver.view.MovingToClientRouteFragment;
import com.cargopull.executor_driver.view.NewPreOrderButtonFragment;
import com.cargopull.executor_driver.view.NewPreOrderFragment;
import com.cargopull.executor_driver.view.OnlineFragment;
import com.cargopull.executor_driver.view.OrderCostDetailsActionsDialogFragment;
import com.cargopull.executor_driver.view.OrderCostDetailsFragment;
import com.cargopull.executor_driver.view.OrderCostDetailsRouteFragment;
import com.cargopull.executor_driver.view.OrderFulfillmentActionsDialogFragment;
import com.cargopull.executor_driver.view.OrderFulfillmentFragment;
import com.cargopull.executor_driver.view.OrderRouteFragment;
import com.cargopull.executor_driver.view.OrdersHistoryHeaderFragment;
import com.cargopull.executor_driver.view.PreOrderConfirmationFragment;
import com.cargopull.executor_driver.view.PreOrderFragment;
import com.cargopull.executor_driver.view.PreOrdersFragment;
import com.cargopull.executor_driver.view.ProfileFragment;
import com.cargopull.executor_driver.view.SelectedPreOrderConfirmationFragment;
import com.cargopull.executor_driver.view.SelectedPreOrderFragment;
import com.cargopull.executor_driver.view.SelectedVehicleFragment;
import com.cargopull.executor_driver.view.SelectedVehicleOptionsFragment;
import com.cargopull.executor_driver.view.ServerConnectionFragment;
import com.cargopull.executor_driver.view.UpcomingPreOrderConfirmationFragment;
import com.cargopull.executor_driver.view.UpcomingPreOrderFragment;
import com.cargopull.executor_driver.view.UpcomingPreOrderNotificationFragment;
import com.cargopull.executor_driver.view.VehicleOptionsFragment;
import com.cargopull.executor_driver.view.WaitingForClientActionsDialogFragment;
import com.cargopull.executor_driver.view.WaitingForClientFragment;
import com.cargopull.executor_driver.view.WaitingForClientRouteFragment;
import com.cargopull.executor_driver.view.auth.LoginFragment;
import com.cargopull.executor_driver.view.auth.PasswordFragment;

public interface AppComponent {

  void inject(Consumer<AppSettingsService> appSettingsServiceConsumer);

  void inject(MainApplication mainApplication);

  void inject(BaseActivity baseActivity);

  void inject(MovingToClientActivity movingToClientActivity);

  void inject(WaitingForClientActivity waitingForClientActivity);

  void inject(OrderFulfillmentActivity orderFulfillmentActivity);

  void inject(OrderCostDetailsActivity orderCostDetailsActivity);

  void inject(MenuActivity menuActivity);

  void inject(OnlineMenuActivity onlineMenuActivity);

  void inject(NightModeActivity nightModeActivity);

  void inject(OnlineActivity onlineActivity);

  void inject(PreOrdersActivity preOrdersActivity);

  void inject(PasswordActivity passwordActivity);

  void inject(FcmService fcmService);

  void inject(LoginFragment loginFragment);

  void inject(PasswordFragment passwordFragment);

  void inject(MapFragment mapFragment);

  void inject(OnlineFragment onlineFragment);

  void inject(GoOnlineFragment goOnlineFragment);

  void inject(ChooseVehicleFragment chooseVehicleFragment);

  void inject(VehicleOptionsFragment vehicleOptionsFragment);

  void inject(SelectedVehicleOptionsFragment vehicleOptionsFragment);

  void inject(SelectedVehicleFragment selectedVehicleFragment);

  void inject(DriverOrderConfirmationFragment driverOrderConfirmationFragment);

  void inject(ClientOrderConfirmationFragment clientOrderConfirmationFragment);

  void inject(ClientOrderConfirmationTimeFragment clientOrderConfirmationTimeFragment);

  void inject(MovingToClientFragment movingToClientFragment);

  void inject(MovingToClientActionsDialogFragment movingToClientActionsDialogFragment);

  void inject(ActionOrderDetailsFragment actionOrderDetailsFragment);

  void inject(MovingToClientRouteFragment movingToClientRouteFragment);

  void inject(WaitingForClientFragment waitingForClientFragment);

  void inject(WaitingForClientActionsDialogFragment waitingForClientActionsDialogFragment);

  void inject(WaitingForClientRouteFragment waitingForClientRouteFragment);

  void inject(OrderFulfillmentFragment orderFulfillmentFragment);

  void inject(OrderFulfillmentActionsDialogFragment orderFulfillmentActionsDialogFragment);

  void inject(OrderRouteFragment orderRouteFragment);

  void inject(CallToClientFragment callToClientFragment);

  void inject(CallToOperatorFragment callToOperatorFragment);

  void inject(CancelOrderDialogFragment cancelOrderDialogFragment);

  void inject(BalanceFragment balanceFragment);

  void inject(BalanceSummaryFragment balanceSummaryFragment);

  void inject(MenuFragment menuFragment);

  void inject(ServerConnectionFragment serverConnectionFragment);

  void inject(OrderCostDetailsFragment orderCostDetailsFragment);

  void inject(OrderCostDetailsRouteFragment orderCostDetailsRouteFragment);

  void inject(OrderCostDetailsActionsDialogFragment orderCostDetailsActionsDialogFragment);

  void inject(ProfileFragment profileFragment);

  void inject(DriverPreOrderBookingFragment driverPreOrderBookingFragment);

  void inject(PreOrdersFragment preOrdersFragment);

  void inject(PreOrderFragment preOrderFragment);

  void inject(SelectedPreOrderFragment selectedPreOrderFragment);

  void inject(SelectedPreOrderConfirmationFragment selectedPreOrderConfirmationFragment);

  void inject(NewPreOrderFragment newPreOrderFragment);

  void inject(NewPreOrderButtonFragment newPreOrderButtonFragment);

  void inject(DriverPreOrderConfirmationFragment driverPreOrderConfirmationFragment);

  void inject(PreOrderConfirmationFragment preOrderConfirmationFragment);

  void inject(UpcomingPreOrderFragment upcomingPreOrderFragment);

  void inject(UpcomingPreOrderConfirmationFragment upcomingPreOrderConfirmationFragment);

  void inject(UpcomingPreOrderNotificationFragment upcomingPreOrderNotificationFragment);

  void inject(GeoEngagementDialogFragment geoEngagementDialogFragment);

  void inject(OrdersHistoryHeaderFragment ordersHistoryHeaderFragment, int offset);
}

File: app/src/main/java/com/cargopull/executor_driver/di/AppComponent.kt
package com.cargopull.executor_driver.di

import android.content.Context
import com.cargopull.executor_driver.application.*
import com.cargopull.executor_driver.backend.settings.AppSettingsService
import com.cargopull.executor_driver.utils.Consumer
import com.cargopull.executor_driver.utils.Releasable
import com.cargopull.executor_driver.utils.TimeUtils
import com.cargopull.executor_driver.utils.TimeUtilsImpl
import com.cargopull.executor_driver.view.*
import com.cargopull.executor_driver.view.auth.LoginFragment
import com.cargopull.executor_driver.view.auth.PasswordFragment

class AppComponent(private val appContext: Context, private val backendComponent: BackendComponent) : Releasable {

    private val timeUtils: TimeUtils by lazy {
        TimeUtilsImpl()
    }
    private val presentationComponent: PresentationComponent by lazy {
        PresentationComponent(backendComponent, timeUtils)
    }
    private val navigationMapper: NavigationMapper by lazy {
        NavigationMapperImpl(backendComponent.ringTonePlayer, backendComponent.shakeItPlayer)
    }

    fun inject(appSettingsServiceConsumer: Consumer<AppSettingsService>) {
        appSettingsServiceConsumer.accept(
                backendComponent.appSettingsService
        )
    }

    fun inject(mainApplication: MainApplication) {
        mainApplication.setServerConnectionViewModel(
                presentationComponent.serverConnectionViewModel
        )
        mainApplication.setBalanceViewModel(
                presentationComponent.balanceViewModel
        )
        mainApplication.setExecutorStateViewModel(
                presentationComponent.executorStateViewModel
        )
        mainApplication.setOrderViewModel(
                presentationComponent.orderViewModel
        )
        mainApplication.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
        mainApplication.setPreOrderViewModel(
                presentationComponent.preOrderViewModel
        )
        mainApplication.setUpcomingPreOrderViewModel(
                presentationComponent.upcomingPreOrderAvailabilityViewModel
        )
        mainApplication.setPreOrdersListViewModel(
                presentationComponent.preOrdersListViewModel
        )
        mainApplication.setOrderCostDetailsViewModel(
                presentationComponent.orderCostDetailsViewModel
        )
        mainApplication.setGeoLocationViewModel(
                presentationComponent.geoLocationViewModel
        )
        mainApplication.setMissedOrderViewModel(
                presentationComponent.missedOrderViewModel
        )
        mainApplication.setUpcomingPreOrderMessageViewModel(
                presentationComponent.upcomingPreOrderMessagesViewModel
        )
        mainApplication.setCancelledOrderViewModel(
                presentationComponent.cancelledOrderViewModel
        )
        mainApplication.setCurrentCostPollingViewModel(
                presentationComponent.currentCostPollingViewModel
        )
        mainApplication.setServerTimeViewModel(
                presentationComponent.serverTimeViewModel
        )
        mainApplication.setNavigationMapper(
                navigationMapper
        )
    }

    fun inject(baseActivity: BaseActivity) {
        baseActivity.setAppSettingsService(
                backendComponent.appSettingsService
        )
        baseActivity.setGeoLocationStateViewModel(
                presentationComponent.getGeoLocationStateViewModel(baseActivity)
        )
        baseActivity.setExecutorStateViewModel(
                presentationComponent.executorStateViewModel
        )
        baseActivity.setUpdateMessageViewModel(
                presentationComponent.updateMessageViewModel
        )
        baseActivity.setAnnouncementViewModel(
                presentationComponent.announcementViewModel
        )
        baseActivity.setServerConnectionViewModel(
                presentationComponent.serverConnectionViewModel
        )
        baseActivity.setServerTimeViewModel(
                presentationComponent.serverTimeViewModel
        )
    }

    fun inject(movingToClientActivity: MovingToClientActivity) {
        movingToClientActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(waitingForClientActivity: WaitingForClientActivity) {
        waitingForClientActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(orderFulfillmentActivity: OrderFulfillmentActivity) {
        orderFulfillmentActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(orderCostDetailsActivity: OrderCostDetailsActivity) {
        orderCostDetailsActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(menuActivity: MenuActivity) {
        menuActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(onlineMenuActivity: OnlineMenuActivity) {
        onlineMenuActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(nightModeActivity: NightModeActivity) {
        nightModeActivity.setAppSettingsService(
                backendComponent.appSettingsService
        )
    }

    fun inject(onlineActivity: OnlineActivity) {
        onlineActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(preOrdersActivity: PreOrdersActivity) {
        preOrdersActivity.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(passwordActivity: PasswordActivity) {
        passwordActivity.setApiService(
                backendComponent.apiService
        )
        passwordActivity.setErrorReporter(
                backendComponent.errorReporter
        )
    }

    fun inject(fcmService: FcmService) {
        fcmService.fcmObserver = backendComponent.fcmReceiver
        fcmService.apiService = backendComponent.apiService
        fcmService.shakeItPlayer = backendComponent.shakeItPlayer
        fcmService.ringTonePlayer = backendComponent.ringTonePlayer
    }

    fun inject(loginFragment: LoginFragment) {
        loginFragment.setAppSettings(
                backendComponent.appSettingsService
        )
        loginFragment.setPhoneViewModel(
                presentationComponent.getPhoneViewModel(loginFragment)
        )
    }

    fun inject(passwordFragment: PasswordFragment) {
        passwordFragment.setSmsButtonViewModel(
                presentationComponent.getSmsButtonViewModel(passwordFragment)
        )
        passwordFragment.setCodeHeaderViewModel(
                presentationComponent.getCodeHeaderViewModel(passwordFragment)
        )
        passwordFragment.setCodeViewModel(
                presentationComponent.getCodeViewModel(passwordFragment)
        )
        passwordFragment.setShakeItPlayer(
                backendComponent.shakeItPlayer
        )
    }

    fun inject(mapFragment: MapFragment) {
        mapFragment.setMapViewModel(
                presentationComponent.getMapViewModel(mapFragment)
        )
        mapFragment.setGeoLocationViewModel(
                presentationComponent.geoLocationViewModel
        )
    }

    fun inject(onlineFragment: OnlineFragment) {
        onlineFragment.setOnlineSwitchViewModel(
                presentationComponent.getOnlineSwitchViewModel(onlineFragment)
        )
        onlineFragment.setOnlineButtonViewModel(
                presentationComponent.getSelectedOnlineButtonViewModel(onlineFragment)
        )
    }

    fun inject(goOnlineFragment: GoOnlineFragment) {
        goOnlineFragment.setOnlineButtonViewModel(
                presentationComponent.getOnlineButtonViewModel(goOnlineFragment)
        )
    }

    fun inject(chooseVehicleFragment: ChooseVehicleFragment) {
        chooseVehicleFragment.setChooseVehicleViewModel(
                presentationComponent.getChooseVehicleViewModel(chooseVehicleFragment)
        )
    }

    fun inject(vehicleOptionsFragment: VehicleOptionsFragment) {
        vehicleOptionsFragment.setVehicleOptionsViewModel(
                presentationComponent.getVehicleOptionsViewModel(vehicleOptionsFragment)
        )
    }

    fun inject(vehicleOptionsFragment: SelectedVehicleOptionsFragment) {
        vehicleOptionsFragment.setVehicleOptionsViewModel(
                presentationComponent.getSelectedVehicleOptionsViewModel(vehicleOptionsFragment)
        )
    }

    fun inject(selectedVehicleFragment: SelectedVehicleFragment) {
        selectedVehicleFragment.setSelectedVehicleViewModel(
                presentationComponent.getSelectedVehicleViewModel(selectedVehicleFragment)
        )
        selectedVehicleFragment.setChooseVehicleViewModel(
                presentationComponent.getSelectedChooseVehicleViewModel(selectedVehicleFragment)
        )
    }

    fun inject(currentVehicleFragment: CurrentVehicleFragment) {
        currentVehicleFragment.setSelectedVehicleViewModel(
                presentationComponent.getSelectedVehicleViewModel(currentVehicleFragment)
        )
        currentVehicleFragment.setChooseVehicleViewModel(
                presentationComponent.getCurrentChooseVehicleViewModel(currentVehicleFragment)
        )
    }

    fun inject(driverOrderConfirmationFragment: DriverOrderConfirmationFragment) {
        driverOrderConfirmationFragment.setOrderConfirmationViewModel(
                presentationComponent.getRushOrderConfirmationViewModel(driverOrderConfirmationFragment)
        )
        driverOrderConfirmationFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
    }

    fun inject(clientOrderConfirmationFragment: ClientOrderConfirmationFragment) {
        clientOrderConfirmationFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
    }

    fun inject(clientOrderConfirmationTimeFragment: ClientOrderConfirmationTimeFragment) {
        clientOrderConfirmationTimeFragment.setClientOrderConfirmationTimeViewModel(
                presentationComponent.getClientOrderConfirmationTimeViewModel(
                        clientOrderConfirmationTimeFragment
                )
        )
    }

    fun inject(movingToClientFragment: MovingToClientFragment) {
        movingToClientFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
        movingToClientFragment.setMovingToClientTimerViewModel(
                presentationComponent.getMovingToClientTimerViewModel(movingToClientFragment)
        )
        movingToClientFragment.setReportArrivedViewModel(
                presentationComponent.getReportArrivedViewModel(movingToClientFragment)
        )
        movingToClientFragment.setShakeItPlayer(
                backendComponent.shakeItPlayer
        )
    }

    fun inject(movingToClientActionsDialogFragment: MovingToClientActionsDialogFragment) {
        movingToClientActionsDialogFragment.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(actionOrderDetailsFragment: ActionOrderDetailsFragment) {
        actionOrderDetailsFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
    }

    fun inject(movingToClientRouteFragment: MovingToClientRouteFragment) {
        movingToClientRouteFragment.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
    }

    fun inject(waitingForClientFragment: WaitingForClientFragment) {
        waitingForClientFragment.setStartOrderViewModel(
                presentationComponent.getStartOrderViewModel(waitingForClientFragment)
        )
        waitingForClientFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
        waitingForClientFragment.setShakeItPlayer(
                backendComponent.shakeItPlayer
        )
    }

    fun inject(waitingForClientActionsDialogFragment: WaitingForClientActionsDialogFragment) {
        waitingForClientActionsDialogFragment.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(waitingForClientRouteFragment: WaitingForClientRouteFragment) {
        waitingForClientRouteFragment.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
    }

    fun inject(orderFulfillmentFragment: OrderFulfillmentFragment) {
        orderFulfillmentFragment.setOrderTimeViewModel(
                presentationComponent.getOrderTimeViewModel(orderFulfillmentFragment)
        )
        orderFulfillmentFragment.setOrderCostViewModel(
                presentationComponent.getOrderCostViewModel(orderFulfillmentFragment)
        )
        orderFulfillmentFragment.setNextRoutePointViewModel(
                presentationComponent.getNextRoutePointViewModel(orderFulfillmentFragment)
        )
        orderFulfillmentFragment.setCompleteOrderViewModel(
                presentationComponent.getCompleteOrderViewModel(orderFulfillmentFragment)
        )
        orderFulfillmentFragment.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
        orderFulfillmentFragment.setShakeItPlayer(
                backendComponent.shakeItPlayer
        )
    }

    fun inject(orderFulfillmentActionsDialogFragment: OrderFulfillmentActionsDialogFragment) {
        orderFulfillmentActionsDialogFragment.setEventLogger(
                backendComponent.eventLogger
        )
        orderFulfillmentActionsDialogFragment.setNextRoutePointViewModel(
                presentationComponent.getNextRoutePointViewModel(orderFulfillmentActionsDialogFragment)
        )
        orderFulfillmentActionsDialogFragment.setCompleteOrderViewModel(
                presentationComponent.getCompleteOrderViewModel(orderFulfillmentActionsDialogFragment)
        )
    }

    fun inject(orderRouteFragment: OrderRouteFragment) {
        orderRouteFragment.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
    }

    fun inject(callToClientFragment: CallToClientFragment) {
        callToClientFragment.setCallToClientViewModel(
                presentationComponent.getCallToClientViewModel(callToClientFragment)
        )
    }

    fun inject(callToOperatorFragment: CallToOperatorFragment) {
        callToOperatorFragment.setCallToOperatorViewModel(
                presentationComponent.getCallToOperatorViewModel(callToOperatorFragment)
        )
    }

    fun inject(cancelOrderDialogFragment: ReportProblemDialogFragment) {
        cancelOrderDialogFragment.setReportProblemViewModel(
                presentationComponent.getCancelOrderViewModel(cancelOrderDialogFragment)
        )
    }

    fun inject(balanceFragment: BalanceFragment) {
        balanceFragment.setBalanceViewModel(
                presentationComponent.balanceViewModel
        )
    }

    fun inject(balanceSummaryFragment: BalanceSummaryFragment) {
        balanceSummaryFragment.setBalanceViewModel(
                presentationComponent.balanceViewModel
        )
    }

    fun inject(menuFragment: MenuFragment) {
        menuFragment.setAppSettingsService(
                backendComponent.appSettingsService
        )
        menuFragment.setBalanceViewModel(
                presentationComponent.balanceViewModel
        )
        menuFragment.setOnlineSwitchViewModel(
                presentationComponent.getExitOnlineSwitchViewModel(menuFragment)
        )
        menuFragment.setPreOrdersListViewModel(
                presentationComponent.preOrdersListViewModel
        )
        menuFragment.setMenuViewModel(
                presentationComponent.getMenuViewModel(menuFragment)
        )
        menuFragment.setOnlineButtonViewModel(
                presentationComponent.getSelectedOnlineButtonViewModel(menuFragment)
        )
    }

    fun inject(serverConnectionFragment: ServerConnectionFragment) {
        serverConnectionFragment.setServerConnectionViewModel(
                presentationComponent.serverConnectionViewModel
        )
    }

    fun inject(orderCostDetailsFragment: OrderCostDetailsFragment) {
        orderCostDetailsFragment.setOrderCostDetailsViewModel(
                presentationComponent.orderCostDetailsViewModel
        )
        orderCostDetailsFragment.setConfirmOrderPaymentViewModel(
                presentationComponent.getConfirmOrderPaymentViewModel(orderCostDetailsFragment)
        )
        orderCostDetailsFragment.setShakeItPlayer(
                backendComponent.shakeItPlayer
        )
    }

    fun inject(orderCostDetailsRouteFragment: OrderCostDetailsRouteFragment) {
        orderCostDetailsRouteFragment.setOrderRouteViewModel(
                presentationComponent.orderRouteViewModel
        )
    }

    fun inject(orderCostDetailsActionsDialogFragment: OrderCostDetailsActionsDialogFragment) {
        orderCostDetailsActionsDialogFragment.setEventLogger(
                backendComponent.eventLogger
        )
    }

    fun inject(profileFragment: ProfileFragment) {
        profileFragment.setAppSettings(
                backendComponent.appSettingsService
        )
    }

    fun inject(driverPreOrderBookingFragment: DriverPreOrderBookingFragment) {
        driverPreOrderBookingFragment.setOrderConfirmationViewModel(
                presentationComponent.getPreOrderBookingViewModel(driverPreOrderBookingFragment)
        )
    }

    fun inject(preOrdersFragment: PreOrdersFragment) {
        preOrdersFragment.setPreOrdersListViewModel(
                presentationComponent.preOrdersListViewModel
        )
    }

    fun inject(preOrderFragment: PreOrderFragment) {
        preOrderFragment.setOrderViewModel(
                presentationComponent.getPOrderViewModel(preOrderFragment)
        )
    }

    fun inject(selectedPreOrderFragment: SelectedPreOrderFragment) {
        selectedPreOrderFragment.setOrderViewModel(
                presentationComponent.getSelectedPreOrderViewModel(selectedPreOrderFragment)
        )
    }

    fun inject(selectedPreOrderConfirmationFragment: SelectedPreOrderConfirmationFragment) {
        selectedPreOrderConfirmationFragment.setOrderConfirmationViewModel(
                presentationComponent
                        .getSelectedPreOrderConfirmationViewModel(selectedPreOrderConfirmationFragment)
        )
    }

    fun inject(newPreOrderFragment: NewPreOrderFragment) {
        newPreOrderFragment.setPreOrderViewModel(
                presentationComponent.preOrderViewModel
        )
    }

    fun inject(newPreOrderButtonFragment: NewPreOrderButtonFragment) {
        newPreOrderButtonFragment.setPreOrderViewModel(
                presentationComponent.preOrderViewModel
        )
    }

    fun inject(driverPreOrderConfirmationFragment: DriverPreOrderConfirmationFragment) {
        driverPreOrderConfirmationFragment.setOrderConfirmationViewModel(
                presentationComponent.getPreOrderConfirmationViewModel(driverPreOrderConfirmationFragment)
        )
        driverPreOrderConfirmationFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
    }

    fun inject(preOrderConfirmationFragment: PreOrderConfirmationFragment) {
        preOrderConfirmationFragment.setOrderViewModel(
                presentationComponent.orderViewModel
        )
    }

    fun inject(upcomingPreOrderFragment: UpcomingPreOrderFragment) {
        upcomingPreOrderFragment.setOrderViewModel(
                presentationComponent.getUpcomingPreOrderViewModel(upcomingPreOrderFragment)
        )
    }

    fun inject(upcomingPreOrderConfirmationFragment: UpcomingPreOrderConfirmationFragment) {
        upcomingPreOrderConfirmationFragment.setOrderConfirmationViewModel(
                presentationComponent
                        .getUpcomingPreOrderConfirmationViewModel(upcomingPreOrderConfirmationFragment)
        )
    }

    fun inject(upcomingPreOrderNotificationFragment: UpcomingPreOrderNotificationFragment) {
        upcomingPreOrderNotificationFragment.setUpcomingPreOrderViewModel(
                presentationComponent
                        .getUpcomingPreOrderViewModel(upcomingPreOrderNotificationFragment)
        )
        upcomingPreOrderNotificationFragment.setUpcomingPreOrderNotificationViewModel(
                presentationComponent.upcomingPreOrderAvailabilityViewModel
        )
    }

    fun inject(geoEngagementDialogFragment: GeoEngagementDialogFragment) {
        geoEngagementDialogFragment.setGeoLocationStateViewModel(
                presentationComponent.getGeoLocationStateViewModel(
                        geoEngagementDialogFragment
                )
        )
    }

    fun inject(ordersHistoryHeaderFragment: OrdersHistoryHeaderFragment, offset: Int) {
        ordersHistoryHeaderFragment.setOrdersHistoryHeaderViewModel(
                presentationComponent
                        .getOrdersHistoryHeaderViewModel(ordersHistoryHeaderFragment, offset)
        )
    }

    override fun release() {
        backendComponent.release()
    }
}
